src/standalone/Vegetation/pmodel.jl

```diff
@@ -337,4 +431,568 @@ function compute_full_pmodel_outputs(
+
+    # drivers 
+    FT = eltype(parameters)
+    # TODO: replace this with modular soil moisture stress parameterization
+    ψ = p.canopy.hydraulics.ψ
+    n = canopy.hydraulics.n_leaf + canopy.hydraulics.n_stem
+    grav = LP.grav(earth_param_set)
+    ρ_water = LP.ρ_cloud_liq(earth_param_set)
+    βm = @. lazy(moisture_stress(ψ.:($$n) * ρ_water * grav, parameters.sc, parameters.pc))
+    T_canopy = canopy_temperature(canopy.energy, canopy, Y, p)
+    VPD = @. lazy(
+        ClimaLand.vapor_pressure_deficit(
+            p.drivers.T,
+            p.drivers.P,
+            p.drivers.q,
+            LP.thermodynamic_parameters(canopy.parameters.earth_param_set),
+        ),
+    )
+    I_abs = @. lazy(
+        compute_I_abs(
+            p.canopy.radiative_transfer.par.abs,
+            p.canopy.radiative_transfer.par_d,
+            canopy.radiative_transfer.parameters.λ_γ_PAR,
+            constants.lightspeed,
+            constants.planck_h,
+            constants.N_a,
+        ),
+    )
+
+    @. p.canopy.photosynthesis.OptVars = update_optimal_EMA(
+        parameters,
+        constants,
+        p.canopy.photosynthesis.OptVars,
+        T_canopy,
+        p.drivers.P,
+        VPD,
+        p.drivers.c_co2,
+        βm,
+        I_abs,
+        local_noon_mask,
+        verbose,
+    )
+end
+
+"""
+function make_PModel_callback(FT, start_date, dt, canopy, longitude=nothing) 
+
+This constructs a FrequencyBasedCallback for the P-model that updates the optimal photosynthetic capacities
+using an exponential moving average at local noon. 
+
+We check for local noon using the provided `longitude` (once passing 
+in lat/lon for point/column domains, this can be automatically extracted from the domain axes) every dt.
+The time of local noon is expressed in seconds UTC and neglects the effects of obliquity and eccentricity, so
+it is constant throughout the year. This presents an error of up to ~20 minutes, but it is sufficient for our 
+application here (since meteorological drivers are often updated at coarser time intervals anyway). 
+
+Args
+- `FT`: The floating-point type used in the model (e.g., `Float32`, `Float64`).
+- `start_date`: datetime object for the start of the simulation (UTC).
+- `dt`: timestep in seconds (this will get cast to type FT)
+- `canopy`: the canopy object containing the P-model parameters and constants.
+- `longitude`: optional longitude in degrees for local noon calculation (default is `nothing`). If we are on 
+    a ClimaLand.Domains.Point, this will need to be supplied explicitly. Otherwise, if we are on a field, then 
+    the longitude at each point can be automatically extracted from the field axes (THIS STILL NEEDS TO BE TESTED)
+"""
+function make_PModel_callback(
+    ::Type{FT},
+    start_date, 
+    dt, 
+    canopy, 
+    longitude = nothing
+) where {FT <: AbstractFloat}
+    function seconds_after_midnight(t)
+        return Hour(t).value * 3600 + Minute(t).value * 60
```

Comment:
add any remaining seconds?
```suggestion
        return Hour(t).value * 3600 + Minute(t).value * 60 + Second(t)
```